# Algorithms Homework 1

In [1]:
# Load required Python libraries.
import matplotlib.pyplot as plt
import numpy as np
from scipy import linalg

## Linear Algebra Review

### Exact solution

Solve the following system of linear equations BY HAND with exact arithmetic:

$$3 x_1 + 2 x_2 + x_3 = 6$$
$$-x_1 + 4 x_2 + 5x_3 = 8$$
$$2x_1 -8 x_2 + 10 x_3 = 4$$

Turn in your work via Gradescope.

### Solve using the inverse

Solve the same linear system by first inverting the matrix A and performing matrix multiplication. You should use Python and SciPy.
* Tutorials and documentation: https://docs.scipy.org/doc/scipy-1.1.0/reference/tutorial/linalg.html

In [2]:
# Add your solution here
A1 = np.array([[3, 2, 1], [-1, 4, 5], [2, -8, 10]])
b = np.array([6, 8, 4])
print("A= \n")
print(A1)
print("=================")
print("b= \n")
print(b)

A= 

[[ 3  2  1]
 [-1  4  5]
 [ 2 -8 10]]
b= 

[6 8 4]


In [3]:
# Add your solution here
A1_inv = np.linalg.inv(A1)
print("Inverse of A is: \n")
print(A1_inv)

Inverse of A is: 

[[ 0.28571429 -0.1         0.02142857]
 [ 0.07142857  0.1        -0.05714286]
 [ 0.          0.1         0.05      ]]


In [4]:
# Add your solution here
solution1=A1_inv@b
print("Solution is (Inverse method): \n")
print(solution1)

Solution is (Inverse method): 

[1. 1. 1.]


### Solve using LU decomposition

Do the following:
1. Use ``linalg.lu(A)`` to calculate $P$, $L$, and $U$.
2. Write a function to solve any linear system given the factorization and $b$. You may not use ```linalg.solve``` in your function. Instead, you should write loops to implement back substitution.
3. Use your function to solve the linear system.


In [5]:
# You need to define the matrix A1 somewhere (or change the variable name)
P, L, U = linalg.lu(A1)

# Permutation matrix
print(P)

# Lower diagonal matrix
print(L)

# Upper diagonal matrix
print(U)

[[1. 0. 0.]
 [0. 0. 1.]
 [0. 1. 0.]]
[[ 1.          0.          0.        ]
 [ 0.66666667  1.          0.        ]
 [-0.33333333 -0.5         1.        ]]
[[ 3.          2.          1.        ]
 [ 0.         -9.33333333  9.33333333]
 [ 0.          0.         10.        ]]


Tip: We discussed this algorithm in class. First write pseudocode on paper (how to translate our notes into loops, logical statements, etc.?). Once you are happy with the logic, code it up.

In [6]:
# Write function here.
def my_lu_solve(P, L, U, b, LOUD=True):
    ''' 
    Solves linear system Ax = b given PLU decomposition of A
    
    Arguments:
        P - N by N permutation matrix
        L - N by N lower triangular matrix with 1 on diagonal
        U - N by N upper triangular matrix
        b - N by 1 vector
    
    Returns:
        x - N by 1 solution to linear system
    '''
    
    # Define x so this does not give an error. You can delete the line below.   
    # Add your solution here
    # Check dimensions
    [Nrow_L, Ncol_L] = L.shape
    assert Nrow_L == Ncol_L
    [Nrow_U, Ncol_U] = U.shape
    assert Nrow_U == Ncol_U
    assert Nrow_U == Nrow_L
    N = Nrow_U
    assert N == b.size
    
    # forward solve
    bmodif = P.dot(b)
    y = np.zeros(N)

    if LOUD:
        print("\nForward solve...")
    for r in range(0,N): # first for loop
        RHS = bmodif[r]
        
        # loop over some of the columns to evaluated summation
        for c in range(0,r):
            RHS -= y[c]*L[r,c]
        
        if LOUD:
            print("y[",r,"]=",RHS)
        
        y[r] = RHS
    
    # backward solve
    x = np.zeros(N)
    if LOUD:
        print("\nBackward solve...")
    for r in range(N-1,-1,-1): # second for loop
        RHS = y[r]
        
        # loop over some of the columns to evaluated summation
        for c in range(r+1,N):
            RHS -= x[c]*U[r,c]
        
        if LOUD:
            print("x[",r,"]=",RHS/U[r,r])
        x[r] = RHS/U[r,r]
    
    return x

In [7]:
x = my_lu_solve(P, L, U, b, LOUD=True)


Forward solve...
y[ 0 ]= 6.0
y[ 1 ]= 0.0
y[ 2 ]= 10.0

Backward solve...
x[ 2 ]= 1.0
x[ 1 ]= 1.0
x[ 0 ]= 1.0


### Solve using `linalg.solve`


In [8]:
# Add your solution here
Solution2 = np.linalg.solve(A1,b)
print("Solution is (Scipy solve method): \n")
print(Solution2)

Solution is (Scipy solve method): 

[1. 1. 1.]


## Eigenvalues

### Calculate eigenvalues by hand (on paper)

Calculate the eigenvalues for the following matrix:

$$
A = \begin{bmatrix} 0 & -4 & -6 \\ -1 & 0 & -3 \\ 1 & 2 & 5 \end{bmatrix}
$$

You may need to do this for a small system on an exam to characterize stationary points of an optimization problem. Hence I would like you to practice at least once on the homework.

### Calculate eigenvalues using `linalg.eig`

Calculate the eigenvalues and corresponding (right) eigenvectors.

In [9]:
# Add your solution here
A2 = np.array([[0, -4, -6], [-1, 0, -3], [1, 2, 5]])
print("A2= \n")
print(A2)

A2= 

[[ 0 -4 -6]
 [-1  0 -3]
 [ 1  2  5]]


In [10]:
# Add your solution here
U, V = np.linalg.eig(A2)
print("Eigenvalues= ", U)
print("Eigenvectors= ", V)

Eigenvalues=  [1. 2. 2.]
Eigenvectors=  [[-0.81649658 -0.0111458   0.90265832]
 [-0.40824829 -0.8302799   0.1523847 ]
 [ 0.40824829  0.5572352  -0.40247591]]


### Definiteness

Based on your calculations above, is this matrix
1. positive definite
2. positive semi definite
3. negative definite
4. negative semi definite
5. indefinite or
6. cannot say without additional calculations?

Briefly comment to justify your answer.

*Note*: In this class, "briefly comment" means write a few sentences, sketch a picture, write an equation or some combination... whichever you feel is necessary to succinctly provide justification.

**Answer**

A matrix is positive semi definite if all eigen values are nonnegative. Here, as all calculated eigen values have such condition, the matrxi is positive semi-definite.

### Singular Value Decomposition

**TODO:** Make this a separate problem (after the assignment is turned in this year).

### SVD calculation using `linalg.svd`

Calculate the singular value decomposition of the following matrix using `linalg.svd`:

$$
A_3 = \begin{bmatrix} 1 & 2 & 0 & -1 \\ 2 & 4 & 10^{-6} & -2 \\ 0 & 2 & -1 & 10^{-3} \end{bmatrix}
$$

In [11]:
# Add your solution here
A3 = np.array([[1, 2, 0, -1], [2, 4, 10**(-6), -2], [0, 2, -1, 10**-3]])
print("A3= \n")
print(A3)

A3= 

[[ 1.e+00  2.e+00  0.e+00 -1.e+00]
 [ 2.e+00  4.e+00  1.e-06 -2.e+00]
 [ 0.e+00  2.e+00 -1.e+00  1.e-03]]


In [12]:
# Add your solution here
U, S, Vh = np.linalg.svd(A3, full_matrices=True)
print("U= ", U, "\n")
print("S= ", S, "\n")
print("Vh= ", Vh, "\n")

U=  [[-4.25830830e-01  1.36631105e-01 -8.94427217e-01]
 [-8.51661650e-01  2.73262656e-01  4.47213544e-01]
 [-3.05516837e-01 -9.52186674e-01  1.91553449e-07]] 

S=  [5.73316009e+00 1.45975216e+00 3.38134101e-07] 

Vh=  [[-3.71375314e-01 -8.49329490e-01  5.32892822e-02  3.71322025e-01]
 [ 4.67994798e-01 -3.68597169e-01  6.52293569e-01 -4.68647091e-01]
 [-3.77573248e-01  3.77856348e-01  7.56090836e-01  3.78139751e-01]
 [ 7.07460026e-01 -3.53376990e-04 -7.08908571e-10  7.06753272e-01]] 



### Condition number
What is the condition number of the matrix? Calculate it two ways:
1. Using SVD results
2. Using ``np.linalg.cond()``

In [13]:
# Add your solution here
Condition_SVD = S[0]/S[-1]
print("Condition Number= ", Condition_SVD)

Condition Number=  16955285.118358795


In [14]:
# Add your solution here
Condition_Scipy = np.linalg.cond(A3)
print("Condition Number= ", Condition_Scipy)

Condition Number=  16955285.11835879


### Linear system

Consider the linear system $A_4 \cdot x = b$ with $b = [1, 0, 3]^T$ and

$$
A_4 = \begin{bmatrix} 1 & 2 & 0 \\ 2 & 4 & 10^{-1} \\ 0 & 2 & -1 \end{bmatrix} ~.
$$

Approximately how much uncertainty $||\Delta b||_2$ can you tolerate if you want the uncertainty $||\Delta x||_2$ to be less than $10^{-2}$?


In [15]:
# Add your solution here
A4 = np.array([[1, 2, 0], [2, 4, 10**-1], [0, 2, -1]])
b = np.array([1, 0, 3])

# Find the solution
solution_x = np.linalg.solve(A4,b)
norm_x = np.linalg.norm(solution_x)
norm_b = np.linalg.norm(b)
cond_number = np.linalg.cond(A4)
delta_x = 10**-2
delta_b = (delta_x / (cond_number * norm_x)) * norm_b
print("Condition number = ", cond_number)
print("The uncertainty can be included is: ", delta_b)

Condition number =  181.9054022877228
The uncertainty can be included is:  6.160694598769966e-06


**Answer**:

The uncertainty can be included is:  6.160694598769966e-06

In [16]:
A5 = np.array([[1, 2, 0], [2, 4, 0], [0, 2, -1]])
U1, S1, Vh1 = np.linalg.svd(A5, full_matrices=True)
S1

array([5.33070426e+00, 1.25840857e+00, 1.48029737e-16])

### Make it singular


Do/answer the following:
1. Propose a change to a single entry in $A_4$ to make it singular.
2. What are the singular values and condition number after this change?
3. What can you say about the solution to $A_4 \cdot x=b$ after the change?

**Answer:**

1. If (row 2, col 3)=$10^{-1}$ turn to 0, then the row two is a cofficient of row1, then the matrix become singular as the determinant get to zero.

2. The singular values are calculated using scipy and the results are 5.33, 1.25, and 0.0. Also, the condition number turn to infinity as the Sn=0.

3. The solution does not exist as the matrix A is not invertible.

## Convexity

Please turn in via Gradescope.

### Determine if the following functions are convex

1. $x^2 + ax + b, ~x \in \mathbb{R}$
2. $x^3, ~x \in \mathbb{R}$
3. $| x |, ~ x \in [ -5, 5]$
4. $x + y, ~ x \in \mathbb{R}, ~ y \in \mathbb{R}$
5. $x \cdot y, ~ x \in \mathbb{R}, ~ y \in \mathbb{R}$
6. $\log(x), ~ x \in (0,1]$

### Prove the following properties

Consider twice differentiable function $f(x): \mathbb{R}^n \rightarrow \mathbb{R}$.

Recall that $f(x)$ is convex on $x \in \mathbb{R}^n$ if and only if

$f(x+p) \geq f(x) + \nabla f(x)^T p$ for all $x, p \in \mathbb{R}^n$


#### PSD implies Convexity

Prove that if $\nabla^2 f(x)$ is positive semidefinite for all $x \in \mathbb{R}^n$, then $f(x)$ must be convex.


#### Convexity implies PSD

Prove that if $f(x)$ is convex then $\nabla^2 f(x)$ must be positive semidefinite.

#### PD implies Strictly Convex

Prove that if $\nabla^2 f(x)$ is positive definite for all $x \in \mathbb{R}^n$, then $f(x)$ must be strictly convex.